Rework spacecraft attitude controls when using keyboard

CaptainSwag101 opened the pull request:
As the title says, this overrides Orbiter's default attitude control handling to implement multiple different attitude/rotation controller deflection values, and changes the default deflection to be just shy of each vehicle's respective trigger point for direct thruster fire, but still commanding maximum proportional rate where applicable.

To get full deflection of the attitude controller and command direct/hardover RCS firing in the CSM/LM respectively, hold the left or right Alt key while commanding rotation. To move the ACA out of detent in the LM without commanding RCS thrust, hold Ctrl and Alt while commanding rotation.

This allows for more intuitive and nuanced control when using the keyboard, and significantly reduces unintentional thruster fire and wasted fuel, which would always occur previously because Orbiter's default behavior commands 100% stick deflection from keyboard inputs. Therefore, procedures which enable the direct RCS/hardover switches can now be followed safely by keyboard users without the side effects of reduced control fidelity and excess fuel usage, or turning off said switches to work around the issue.

Furthermore, this PR fixes an error that was discovered in the CSM RHC simulation: The direct RCS switches were implemented, but their state was not being checked by the appropriate RCS code, which was instead directly checking the input deflection value for an incorrect deflection threshold of 10.5° instead of 11°.

This PR primarily affects the keyboard control method. However, VESIM and Joystick controls will be affected by the change to the CSM RHC proportional rate scaling.

Some related info: the direct coil trigger point is 11° deflection out of a maximum 11.5° in the CSM, and 12° deflection out of a maximum 13° in the LM, per each vehicle's AOH Volume 1.

CaptainSwag101 committed:
Rework LM attitude keys to not command hardover unless CTRL is pressed

CaptainSwag101 committed:
Rename the deflection variable to indicate it's only for keyboard
Also add a comment explaining how the 0.923 value was derived.

CaptainSwag101 committed:
Don't override DSKY/DEDA shortcuts with ACA inputs

CaptainSwag101 committed:
Fix control inversion, only override Orbiter keys when in Rotation control mode

CaptainSwag101 committed:
Implement non-direct-coil attitude control, fix unused RHC direct switches
The RHC class has its own implementation and checks for the direct coil switches, but mysteriously they went completely unused, and the deflection state was manually checked in the satsystems file, and using a slightly different deflection value than the RHC. I am under the strong assumption that this was a mistake/oversight, and have used the hardstop switches of the RHC now to trigger the direct coils. Everything seems to behave nicely.

CaptainSwag101 committed:
Disable LM and CSM RHC debug string

CaptainSwag101 commented:
I can't make this change from mobile at the moment, but please temporarily consider this PR unfinished. I want to expand its scope to enable the out-of-detent signal for the LM and potentially CSM, to allow for the on-surface thruster checks without using fuel.

n7275 commented:
> I can't make this change from mobile at the moment, but please temporarily consider this PR unfinished. I want to expand its scope to enable the out-of-detent signal for the LM and potentially CSM, to allow for the on-surface thruster checks without using fuel.

I marked it as draft for you.

CaptainSwag101 committed:
Out-of-detent and breakout attitude keyboard support
Also stop the DEDA and DSKY keys from getting simultaneous inputs from the numpad if they aren't getting ONLY the modifier key they specifically want. Previously they only cared if that key was pressed, not that other keys weren't pressed too.

CaptainSwag101 committed:
Fix CSM RHC proportional rate scaling
This previously scaled from 0 to 11.5 degrees, which is wrong. Maximum rate is achieved at 10 degrees of deflection. Minimum rate scaling isn't clear based on the documentation, but for joystick deadband purposes I've set this to 1 degree, the same as the LM.

CaptainSwag101 committed:
Update comments with degree symbols, adjust CSM minimum deflection value

CaptainSwag101 commented:
This PR is now ready for review! I have significantly revised the description above to detail all the changes made. In particular I'd really appreciate some help looking into the minimum value for the CSM rotation hand controller proportional rate transducers. In the LM, proportional rate goes from 1 degree of deflection to 11 degrees. In the CSM it goes from ??? degrees to 10 degrees. The AOH volume 1 makes mention of the rate command signal being routed "through the RC breakout logic" but the schematics it references don't make it clear whether or not the breakout switches, located at 1.5 degrees of deflection, are actually the trigger for proportional rate to begin.

CaptainSwag101 committed:
Correct RHC scaling to begin at 1.5 degrees
Also remove an old and incorrect comment about the RHC from satsystems.cpp

CaptainSwag101 committed:
Add some comments to SCS logic, create a helpful constant

CaptainSwag101 committed:
Change RHC proportional scaling for a wider useful range
Also remove the questioning comments from the SCS code in the previous commit since they aren't truly relevant to this PR.

CaptainSwag101 committed:
Disable RHC debug string

ThymoNL commented:
How are we going to communicate this to new and existing users?
I think it's good to make an article on the wiki about this and some sort of announcement on the forum also wouldn't hurt.

CaptainSwag101 commented:
I definitely plan on making a forum post about this. My hope, also, is that this change will be mostly not noticed by users if they don't happen to see it mentioned. The default behavior under most circumstances will remain unchanged; only in edge cases such as engine burns will the behavior differ, and even then the behavior ought to be more desirable than it was before. I think the "NASSP keyboard controls" forum post also ought to be updated with this new information.

ThatRedMelon commented on Orbitersdk/samples/ProjectApollo/src_csm/scs.cpp:
```diff
@@ -4106,9 +4106,9 @@ void ECA::TimeStep(double simdt) {
-		// sprintf(oapiDebugString(),"SCS: RATE CMD r%.3f p%.3f y%.3f ERR r%.3f p%.3f y%.3f",
-		//	cmd_rate.x * DEG, cmd_rate.y * DEG, cmd_rate.z * DEG, 
-		//	rate_err.x * DEG, rate_err.y * DEG, rate_err.z * DEG);	
+		 sprintf(oapiDebugString(),"SCS: RATE CMD r%.3f p%.3f y%.3f ERR r%.3f p%.3f y%.3f",
```
Should this still be uncommented?

CaptainSwag101 replied:
No it should not, nice catch!

ThatRedMelon reviewed (commented):
Overall the entire PR works with the correct keybinding and expected results. I can't test the vesim or joystick implementation as I don't have one to try, but the numpad controls work perfectly. Seems all good on my end, but I'll save on approving until someone can test with a joystick/vesim, as well as the debug line I left a comment on.

CaptainSwag101 committed:
Disable debug string

indy91 reviewed (changes requested):
Just the initialization of the rhc_keyboard_deflection variables as we talked about on Discord, then this PR is ready to go!

CaptainSwag101 committed:
Initialize keyboard deflection arrays at vessel init

CaptainSwag101 committed:
Revert RHC proportional rate changes
Seems likely that trying to imitate the LM ACA behavior is incorrect... no evidence to suggest the CM RHC scales its rate that way.
